test/integration/servicecidr/migration_test.go

```diff
@@ -0,0 +1,292 @@
+		cidr, err := client1.NetworkingV1alpha1().ServiceCIDRs().Get(context.TODO(), defaultservicecidr.DefaultServiceCIDRName, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return false, err
+		}
+		for _, condition := range cidr.Status.Conditions {
+			if condition.Type == networkingv1alpha1.ServiceCIDRConditionReady {
+				return condition.Status == metav1.ConditionFalse, nil
+			}
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("waiting for default service cidr ready condition set to false")
+	}
+
+	// Migrate the services, delete the existing ones and recreate without specifying the ClusterIP
+	services, err := client1.CoreV1().Services("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, svc := range services.Items {
+		// skip the default service since is managed by the apiserver
+		// and we want the new apiserver with the new service cidr to take over
+		if svc.Name == "kubernetes" {
+			continue
+		}
+		// wipe the necessary fields so we can recreate the Service
+		svc.ResourceVersion = ""
+		svc.Spec.ClusterIP = ""
+		svc.Spec.ClusterIPs = nil
+		svc.Status = v1.ServiceStatus{}
+		if err := client1.CoreV1().Services(svc.Namespace).Delete(context.Background(), svc.Name, metav1.DeleteOptions{}); err != nil {
+			t.Fatalf("got unexpected error: %v", err)
+		}
+		svc, err := client1.CoreV1().Services(svc.Namespace).Create(context.Background(), &svc, metav1.CreateOptions{})
+		if err != nil {
+			t.Fatalf("got unexpected error: %v", err)
+		}
+		if !strings.Contains(svc.Spec.ClusterIP, "10.168.") {
+			t.Fatalf("Service expected to have an ip in range 10.168.0.0/24, got %s", svc.Spec.ClusterIP)
+		}
+	}
+
+	// start second apiserver with the new range and new service cidr controller
+	s2 := kubeapiservertesting.StartTestServerOrDie(t,
+		apiServerOptions,
+		[]string{
+			"--runtime-config=networking.k8s.io/v1alpha1=true",
+			"--service-cluster-ip-range=" + cidr2,
+			"--advertise-address=10.1.1.1",
+			"--disable-admission-plugins=ServiceAccount",
+		},
+		etcdOptions)
+	defer s2.TearDownFn()
+
+	client2, err := clientset.NewForConfig(s2.ClientConfig)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	defer framework.DeleteNamespaceOrDie(client2, ns, t)
+
+	// switch the controller to the new apiserver
+	cancelFn()
+	s1.TearDownFn()
+
+	// ServiceCIDR controller
+	ctx2, cancelFn2 := context.WithCancel(context.Background())
+	defer cancelFn2()
+	informers2 := informers.NewSharedInformerFactory(client2, resyncPeriod)
+	go servicecidrs.NewController(
+		informers2.Networking().V1alpha1().ServiceCIDRs(),
+		informers2.Networking().V1alpha1().IPAddresses(),
+		client2,
+	).Run(ctx2, 5)
+	informers2.Start(ctx2.Done())
+
+	// delete the kubernetes.default service so the old DefaultServiceCIDR can be deleted
+	// and the new apiserver can take over
```

Comment:
>  we could go an arbitrary amount of time without having a default ServiceCIDR.

There will be always a default ServiceCIDR, the thing is that if some admin wants to delete a ServiceCIDR, it has to create an overlapping ServiceCIDR so there are no orphan IPs. In this test, I choose a really complex scenario, sowe completely renumber the default ServiceCIDR and starts with a new range, that is why you need to delete it, but it can not be deleted because the kubernetes.default Service has an IP in that range.

> It seems like it would be better if the autocreated ServiceCIDR instead just had generateName: "kubernetes-" (or "default-" or whatever) rather than a specific hardcoded name, and there was a label or annotation to mark the ServiceCIDR that was currently the default...


That was the initial proposal from Tim, I went ahead and tried to implement it but I ended it discarding it and pushing back because it adds a lot of complexity in a critical path like the apiserver bootstrap and is also a problem that already exist,right now you can run apiserver with different flags and both will start but your Services will be complete broken https://github.com/kubernetes/kubernetes/issues/114743 . It is a kind of `sudo rm -rf /` operation too,

Reply:
> > we could go an arbitrary amount of time without having a default ServiceCIDR.
> 
> There will be always a default ServiceCIDR

Well, there will always be a ServiceCIDR with the `DefaultServiceCIDRName`. But that ServiceCIDR doesn't always represent the `kube-apiserver --service-cluster-ip-range` value. And sometimes, _no_ ServiceCIDR represents the `kube-apiserver --service-cluster-ip-range` value.

Reply:
yes, that is on purpose, so we can make the behavior deterministic and match the existing for the "kubernetes.default" Service.
The unique name provides consistency and we define the object as the source of truth and the flags as the generator of the source of truth, we still need to flesh out more details before going to beta